src/entt/entity/polymorphic.hpp

```diff
@@ -0,0 +1,500 @@
+
+public:
+    using entity_type = typename PoolHolderType::entity_type;
+    using pointer = std::conditional_t<std::is_const_v<PoolHolderType>, typename PoolHolderType::const_pointer_type, typename PoolHolderType::pointer_type>;
+    using value_type = std::conditional_t<std::is_pointer_v<typename PoolHolderType::value_type>, pointer, constness_as_t<typename PoolHolderType::value_type, PoolHolderType>>;
+    using reference = std::conditional_t<std::is_pointer_v<typename PoolHolderType::value_type>, pointer, value_type&>;
+    using difference_type = typename iterator_traits::difference_type;
+    using iterator_category = std::forward_iterator_tag;
+
+    poly_components_iterator() ENTT_NOEXCEPT = default;
+
+    poly_components_iterator(entity_type e, PoolsIterator pos, PoolsIterator last) ENTT_NOEXCEPT :
+        ent(e), it(pos), end(last) {
+        if (it != end) {
+            current = static_cast<PoolHolderType&>(*it).try_get(ent);
+            if(current == nullptr) {
+                ++(*this);
+            }
+        }
+    }
+
+    poly_components_iterator &operator++() ENTT_NOEXCEPT {
+        ++it;
+        while (it != end) {
+            current = static_cast<PoolHolderType&>(*it).try_get(ent);
+            if (current != nullptr)
+                break;
+            ++it;
+        }
+        return *this;
+    }
+
+    poly_components_iterator operator++(int) ENTT_NOEXCEPT {
+        poly_components_iterator orig = *this;
+        return ++(*this), orig;
+    }
+
+    bool operator==(const poly_components_iterator& other) const ENTT_NOEXCEPT {
+        return it == other.it;
+    }
+
+    bool operator!=(const poly_components_iterator& other) const ENTT_NOEXCEPT {
+        return it != other.it;
+    }
+
+    pointer operator->() ENTT_NOEXCEPT {
+        return current;
+    }
+
+    reference operator*() ENTT_NOEXCEPT {
+        if constexpr(std::is_pointer_v<value_type>) {
+            return operator->();
+        } else {
+            return *operator->();
+        }
+    }
+
+private:
+    entity_type ent;
+    PoolsIterator it;
+    PoolsIterator end;
+    pointer current = nullptr;
+};
+
+} // namespace entt::internal
+
+/**
+ * Internal details not to be documented.
+ * @endcond
+ */
+
+
+namespace entt {
+
+/** @brief base class for entt::poly_pool_holder */
+template<typename Entity>
+class poly_pool_holder_base {
+public:
+    inline poly_pool_holder_base(void* pool,
+                                 void* (*getter)(void*, Entity) ENTT_NOEXCEPT) :
+          pool_ptr(pool), getter_ptr(getter) {}
+
+protected:
+    void* pool_ptr;
+    void* (*getter_ptr)(void*, Entity) ENTT_NOEXCEPT;
+};
+
+/**
+ * @brief Holds pointer to child type storage and a function to convert it into parent type
+ * @tparam Entity underlying entity type
+ * @tparam Type polymorphic component type to convert into
+ * @tparam Allocator allocator type of the pool
+ */
+template<typename Entity, typename Type, typename Allocator>
```

Comment:
Isn't this the same as using `type Storage`, then taking `typename Storage::entity_type` and the others (ie `value_type` and `allocator_type`) from it?
It would also vastly simplify the definition at a first glance.

Reply:
Yeah, but `poly_type` has to be created for the types, that could not have storage type (most likely ones with pure virtual methods). The possible workaround here, that I see right away, is to create `storage_traits` specialization for polymorphic types, that cannot have a storage type, with some placeholder storage type, that only provides `entity_type`, `value_type` and `allocator_type`. Then, the `storage_traits` could be used to access polymorphic types by pair of entity and component type.

But this look like something, that could possibly break something in the abstraction of polymorphic algorithms from registry, that I cannot see right away. Also, this will not simplify things for the end user at all, because he would still be required to declare custom allocator type in any case. And either it will not make life a lot simpler inside the implementation, because all types for `poly_type`, except the component type itself, is acquired from the `poly_types_accessor` anyway.